Initial integration of the Unitxt to LM eval harness

yoavkatz opened the pull request:
This PR is an initial integration of the [Unitxt](https://github.com/IBM/unitxt) data processing and evaluation library into LM-eval harness.

Unitxt supports a variety of NLP tasks, datasets, templates and metrics.
The Unitxt catalog can be  viewed in an [online explorer](https://unitxt.readthedocs.io/en/latest/docs/demo.html).
An overview of Unitxt can be found in the [paper](https://arxiv.org/abs/2401.14019).

This integration does require any code changes in the core of LM eval harness.
There are only 4 files to be reviewed (generate_yamls.py, template.yaml.file, unitxt_datasets,  unitxt_wrapper.py)
 and the rest are auto generated.

The  **generate_yamls.py** script reads the **unitxt_datasets** file that contains a list of Unitxt datasets and 
creates a corresponding LM-eval YAML for each Unitxt dataset.   Only datasets that do not already exist in LM eval harness were added.

The output YAML is based on **template.yaml.file** and is modified to include the Unitxt card and metric specifications..

In addition, there is a shared python  file **unitxt_wrapper.py** (40 lines)  with functions that are called in the YAML files and that register Unitxt metrics to lm-eval.

Running  lm-eval on a Unitxt dataset Is done as for any other task.

```
lm_eval --model hf --model_args pretrained=google/flan-t5-base 
    --tasks ledgar --device cuda:0 
    --batch_size 8 --limit 100 --output_path output --log_samples
```

Considerations:
1. Unitxt supports dynamic configuration of templates (prompts),  Currently we generate one YAML per dataset with one specific default template.
2. Unitxt supports dynamic configuration of system prompt, format, and number of demonstrations.  Currently, the are assumed to be handled by LM eval harness.
3. Tighter integration of Unitxt and LM-harness could remove the need for generating the static yamls and allow direct access to to Unitxt data card definitions.

```
lm_eval --model hf --model_args pretrained=google/flan-t5-base 
    --tasks unitext:card=cards.ag_news,template=templates.classification.multi_class.default
   --device cuda:0 
    --batch_size 8 --limit 100 --output_path output --log_samples
 ```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=1615) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=1615) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EleutherAI/lm-evaluation-harness?pullRequest=1615) it.</sub>

lintangsutawika commented on lm_eval/tasks/unitxt/tasks.classification.multi_class.yaml:
```diff
@@ -0,0 +1,26 @@
+group:
+- unitxt
+task: TASK_PLACEHOLDER
```
Task templates don't require placeholders, but it may help a user to remind them what needs to be filled in if they want to include this template. I think `task` and `dataset_name` could be commented out so users can see them.

lintangsutawika commented on lm_eval/tasks/unitxt/tasks.classification.multi_class.yaml:
```diff
@@ -0,0 +1,26 @@
+group:
+- unitxt
+task: TASK_PLACEHOLDER
+dataset_path: unitxt/data
+dataset_name: card=CARD_PLACEHOLDER,template=TEMPLATE_PLACEHOLDER
+output_type: generate_until
+training_split: train
+validation_split: test
+doc_to_text: '{{source}}'
+doc_to_target: target
+process_results: !function 'unitxt_wrapper.process_results'
+generation_kwargs:
+  until:
+  - </s>
+metric_list:
+- metric: unitxt_f1_micro
+  aggregation: unitxt
+  higher_is_better: true
+- metric: unitxt_accuracy
+  aggregation: unitxt
+  higher_is_better: true
+- metric: unitxt_f1_macro
+  aggregation: unitxt
+  higher_is_better: true
+metadata:
+  version: 1.0
```
Should be rename to something like 
`_tasks_classification_multi_class_yaml`. Since it's a template, not naming it with `.yaml` file extension will avoid it from being read as a task.

yoavkatz replied:
Sure. I will rename all the files.

lintangsutawika commented on lm_eval/tasks/unitxt/tasks.classification.multi_label.yaml:
```diff
@@ -0,0 +1,26 @@
+group:
+- unitxt
+task: TASK_PLACEHOLDER
+dataset_path: unitxt/data
+dataset_name: card=CARD_PLACEHOLDER,template=TEMPLATE_PLACEHOLDER
+output_type: generate_until
+training_split: train
+validation_split: test
+doc_to_text: '{{source}}'
+doc_to_target: target
+process_results: !function 'unitxt_wrapper.process_results'
+generation_kwargs:
+  until:
+  - </s>
+metric_list:
+- metric: unitxt_f1_micro_multi_label
+  aggregation: unitxt
+  higher_is_better: true
+- metric: unitxt_accuracy
+  aggregation: unitxt
+  higher_is_better: true
+- metric: unitxt_f1_macro_multi_label
+  aggregation: unitxt
+  higher_is_better: true
+metadata:
+  version: 1.0
```
Same comment regarding file naming as `tasks.classification.multi_class.yaml`

lintangsutawika commented on lm_eval/tasks/unitxt/tasks.qa.with_context.extractive.yaml:
```diff
@@ -0,0 +1,20 @@
+group:
+- unitxt
+task: TASK_PLACEHOLDER
+dataset_path: unitxt/data
+dataset_name: card=CARD_PLACEHOLDER,template=TEMPLATE_PLACEHOLDER
+output_type: generate_until
+training_split: train
+validation_split: test
+doc_to_text: '{{source}}'
+doc_to_target: target
+process_results: !function 'unitxt_wrapper.process_results'
+generation_kwargs:
+  until:
+  - </s>
+metric_list:
+- metric: unitxt_squad
+  aggregation: unitxt
+  higher_is_better: true
+metadata:
+  version: 1.0
```
Same comment regarding file naming as `tasks.classification.multi_class.yaml`

lintangsutawika commented on lm_eval/tasks/unitxt/tasks.regression.two_texts.yaml:
```diff
@@ -0,0 +1,20 @@
+group:
+- unitxt
+task: TASK_PLACEHOLDER
+dataset_path: unitxt/data
+dataset_name: card=CARD_PLACEHOLDER,template=TEMPLATE_PLACEHOLDER
+output_type: generate_until
+training_split: train
+validation_split: test
+doc_to_text: '{{source}}'
+doc_to_target: target
+process_results: !function 'unitxt_wrapper.process_results'
+generation_kwargs:
+  until:
+  - </s>
+metric_list:
+- metric: unitxt_spearman
+  aggregation: unitxt
+  higher_is_better: true
+metadata:
+  version: 1.0
```
Same comment regarding file naming as `tasks.classification.multi_class.yaml`

lintangsutawika commented on lm_eval/tasks/unitxt/tasks.summarization.abstractive.yaml:
```diff
@@ -0,0 +1,20 @@
+group:
+- unitxt
+task: TASK_PLACEHOLDER
+dataset_path: unitxt/data
+dataset_name: card=CARD_PLACEHOLDER,template=TEMPLATE_PLACEHOLDER
+output_type: generate_until
+training_split: train
+validation_split: test
+doc_to_text: '{{source}}'
+doc_to_target: target
+process_results: !function 'unitxt_wrapper.process_results'
+generation_kwargs:
+  until:
+  - </s>
+metric_list:
+- metric: unitxt_rouge
+  aggregation: unitxt
+  higher_is_better: true
+metadata:
+  version: 1.0
```
Same comment regarding file naming as `tasks.classification.multi_class.yaml`

lintangsutawika commented on lm_eval/tasks/unitxt/template.yaml.file:
```diff
@@ -0,0 +1,20 @@
+group: 
+  - unitxt
+task: TASK_PLACEHOLDER
+dataset_path: unitxt/data
+dataset_name: card=CARD_PLACEHOLDER,template=TEMPLATE_PLACEHOLDER
+output_type:  generate_until
+training_split: train
+validation_split: test
+doc_to_text: "{{source}}"
+doc_to_target: target
+process_results: !function unitxt_wrapper.process_results
+generation_kwargs:
+  until:
+    - "</s>"
+metric_list:
+  - metric: METRIC_PLACEHOLDER
+    aggregation: unitxt
+    higher_is_better: true
+metadata:
+  version: 1.0
```
Is this template used by the tasks here?

yoavkatz replied:
This is used to generate the  tasks.classification.multi_class.yaml  and other yamls by the generate_yamls script.   I think I can also embed it in the script to avoid the extra file.

lintangsutawika reviewed (commented):
Thanks for the PR!

In addition to the comments I made,

- you may also want to add unitxt dependancies to `pyproject.toml` that users can install it through `pip install lm_eval[unitxt]`
- Add some dependency check in `lm_eval/tasks/unitxt/unitxt_wrapper.py` which will communicate an error/warning if unitxt is not installed yet.

yoavkatz commented:
> > Thanks for the PR!
> > In addition to the comments I made,
> > 
> > * you may also want to add unitxt dependancies to `pyproject.toml` that users can install it through `pip install lm_eval[unitxt]`
> > * Add some dependency check in `lm_eval/tasks/unitxt/unitxt_wrapper.py` which will communicate an error/warning if unitxt is not installed yet.
 
 
 I completed all the changes above. Can you please review and let me know if you have additional feedback.

I also added README and added two new tasks (Entity extraction and grammatical error corrections)

yoavkatz commented:
Hi @lintangsutawika @haileyschoelkopf - I updated the documentation:

https://github.com/EleutherAI/lm-evaluation-harness/pull/1615/files#diff-556af2425890013f9dd8057bc56bb15013d24813bb65981374bf542438156e68

and the pyproject.toml file.
